### DISCUSSION TITLE
Does it make sense to reduce amount of goroutines?

### DISCUSSION THREAD

**Discussion by AskAlexSharov** [anacrolix/torrent#762] [Ideas]
Now it creating: amountOfFiles*amountOfTorrents goroutines to scrape trackers

For example I have system with 87 files (we will have 200 soon) with 7 trackers in each torrent (same 7 trackers), and it producing next amount of goroutines:
```
goroutine profile: total 2685
957 @ 0x4038876 0x40482b2 0x49284ea 0x4069721
#	0x49284e9	github.com/anacrolix/torrent.(*trackerScraper).Run.func1+0xc9	github.com/anacrolix/torrent@v1.44.0/tracker_scraper.go:205

873 @ 0x4038876 0x40482b2 0x492839c 0x4069721
#	0x492839b	github.com/anacrolix/torrent.(*trackerScraper).Run+0x49b	github.com/anacrolix/torrent@v1.44.0/tracker_scraper.go:247
```

It's not a big deal from performance point of view. 
But, even if App doesn't have any load (no seed, no download, no work), it still has unusable big trace (for example if need find deadlock in app). 

Also if I will increase amount of trackers from 7 to 8, it will add 200 new goroutines to my App (smells wrong).

**Comment by anacrolix** [comment]
I'm open to an alternative, but I'm not sure there's a nice one that provides the same level of clean synchronization than the goroutines. They likely require timers and scans and such, and this design is to allow announces to occur synchronously with changes in client state. I have a few observations:

1. I think Go 1.19 removes the race detector goroutine limit, so this kind of thing is less of an issue for testing.
2. There are tools that will scan your goroutine traces and find the cause of issues for you (the numbers here are large, but I don't think they add to the complexity unless they were all in different states).
3. The overhead of having these goroutines should be very minimal.

Let me know what you think.
